### ISSUE TITLE
Extracting Values from Layers for Cost Functions

### ISSUE THREAD

**Issue by willchoprojects** [lululxvi/deepxde#460]
Hello Professor Lu,

 
I've tried to find documentation or Q&A related to my problem, but I couldn't quite find what I'm looking for – I was wondering if you could help me or point me to documentation I could read about this.

I'm trying to implement a cost function that can use the values of one of the layers of the neural network.

For example, if I had a network such as this:
<img width="675" alt="Screen Shot 2021-12-22 at 2 07 17 PM" src="https://user-images.githubusercontent.com/61395682/147143025-21f9f2a5-92df-4a63-b17d-c045cfc169aa.png">

I would like to be able to take the values of the green-boxed layer and use them in my cost function.

Would you have an idea of how to go about this?

Thank you so much for your time, and the amazing documentation!!

William

**Comment by lululxvi** [comment]
I didn't fully understand your question. What do you mean "use them in my cost function"?

**Comment by willchoprojects** [comment]
Hmm for example, if I the system:
<img width="675" alt="Screen Shot 2021-12-22 at 2 07 17 PM" src="https://user-images.githubusercontent.com/61395682/147178338-53d08933-553d-4c67-a2da-be94259eef02.png">

Maybe my cost function could look something like this:
<img width="1075" alt="Screen Shot 2021-12-22 at 9 33 52 PM" src="https://user-images.githubusercontent.com/61395682/147179074-ab1d1cb2-f526-455c-a7aa-2d99aa548147.png">

I could try to come up with a more concrete example if that would be helpful!

Thank you!

William

**Comment by lululxvi** [comment]
Yes, I am still confused about what you are trying to do.

**Comment by willchoprojects** [comment]
Hello again!

Here's my attempt at an example:

<img width="1130" alt="Screen Shot 2021-12-26 at 3 02 55 PM" src="https://user-images.githubusercontent.com/61395682/147418853-8ae5c167-060d-43e8-a995-374334cf802f.png">

And the loss could be calculated as the following:
```
L = [values in green hidden layer] + [L2 loss on outputs]
L = [ |3.1| + |2.0| + |3.3| ] + [ (3.2 - 4.0)^2 + (1.3 - 1.5)^2 ]
```

**Comment by lululxvi** [comment]
Yes, it is doable, but you have to modify the source code.
- You need to modify the network to store the values of the green nodes.
- Then add them into the loss function.
